### PR TITLE
adding fuzzer for  espeak_synth

### PR DESCRIPTION
**Pull request by kmamadoudram**
Hi all,

We are 3 students in IT security (@annastan, @kmamadoudram, @yocvito) and are currently working on fuzzing liblouis/espeak-ng for a school project with @sthibaul. For espeak, , we have decided to focus on espeak_Synth function and have achieved to find known bugs ([here](https://github.com/espeak-ng/espeak-ng/issues/1073)) in previous commits, thus we come here to show you the work we have done.

**So, what have we done yet ?**
Basically, we have used _libfuzzer_ to implements fuzzing on `espeak_Synth`. The fuzzer just look for an ENV variable containing the language you want to test and start fuzzing.
If you are already familiar with libfuzzer, then there is no other specific switch than libfuzzer ones to pass to the fuzzer.
The only thing you need to care about is configuration but you can find more information on how to use the fuzzer we added in [tests/fuzzing/README.md](https://github.com/yocvito/espeak-ng/tree/master/tests/fuzzing#readme)
To check that our fuzzer was actually working, we have used clang coverage and have added the related compiler flags to the espeak-ng directory sources in makefile.am. We were thus able to see that our fuzzer was, indeed, really reaching a lot of the source code. (more info on coverage in the README.md)

**What’s now ?**
We will continue to integrate and improve the fuzzers in the next months. It's not entirely ready yet but we will continue to modify it and come back to you to highlights our changes.
Finally, to find bugs, we need relevant corpus files for each language. We have built a first skeleton of what could be a fuzzer for this functions and have added some scripts to help using it. We are still working on it and will be grateful about any suggestions you could make.

Thank you.

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -942,14 +1022,57 @@ dictsource/yue_emoji:
...
+	espeak-ng-data/mbrola_ph/us_phtrans \
+	espeak-ng-data/mbrola_ph/us3_phtrans \
+	espeak-ng-data/mbrola_ph/vz_phtrans
 
 espeak-ng-data/mbrola_ph/%_phtrans: phsource/mbrola/% src/espeak-ng
 	mkdir -p espeak-ng-data/mbrola_ph
-	ESPEAK_DATA_PATH=$(CURDIR) src/espeak-ng --compile-mbrola=phsource/mbrola/$*
-
-EXTRA_DIST += $(PHSOURCE_MBROLA)
+	ESPEAK_DATA_PATH=$(CURDIR) src/espeak-ng --compile-mbrola=phsource/mbrola/$*
```
Comment (sthibaul) [inline Makefile.am:1078]: You have unrelated changes reversions here. You probably mis-merged master at some point,  we can have a look at that tomorrow

File: Makefile.am [hunk 2]
```diff
@@ -333,21 +313,27 @@ tests_readclause_test_SOURCES += tests/dummy.cpp
...
+tests_ssml_fuzzer_SOURCES = tests/ssml-fuzzer.c
+tests_ssml_fuzzer_LDADD =  src/libespeak-ng.la
+tests_ssml_fuzzer_CFLAGS= -fsanitize=address,undefined,fuzzer -fno-omit-frame-pointer -g -fsanitize-coverage=edge -Isrc/libespeak-ng ${AM_CFLAGS}
+tests_ssml_fuzzer_LDFLAGS= -fsanitize=address,undefined,fuzzer -static -lm  -fsanitize-coverage=edge ${PCAUDIOLIB_LIBS}
+
 if HAVE_LIBFUZZER
-tests_ssml_fuzzer_test_CFLAGS += -fsanitize=fuzzer
-tests_ssml_fuzzer_test_LDFLAGS = -fsanitize=fuzzer
+bin_PROGRAMS +=  tests/fuzzing/synth_fuzzer
+bin_PROGRAMS +=  tests/ssml_fuzzer
```
Comment (sthibaul) [inline Makefile.am:328]: This should rather be `noinst_PROGRAMS`, since these are not tools for end users

File: configure.ac [hunk 3]
```diff
@@ -332,7 +336,19 @@ else
...
+dnl clang-coverage checks.
+dnl ================================================================
+
+if test "$with_coverage" = "yes" ; then
+	use_coverage=yes
+else
+	use_coverage=no
+fi
+
+AM_CONDITIONAL(USE_COVERAGE, [test x"$use_coverage" = xyes])
```
Comment (sthibaul) [inline configure.ac:351]: Directly use $with_coverage

File: tests/fuzzing/synth_fuzzer.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libgen.h>
+#include <time.h>
+
+#include <espeak-ng/espeak_ng.h>
+#define BOLDRED(x) "\x1b[31m\x1b[1m" x "\x1b[0m"
+/***** CONFIG *****/
+#define ESPEAK_DATA_PATH "/usr/lib/x86_64-linux-gnu/espeak-ng-data/"
```
Comment (sthibaul) [inline tests/fuzzing/synth_fuzzer.c:30]: Don't hardcode like this.  Include config.h which defines it

File: Makefile.am (hunk 1 above)
Comment (kmamadoudram) [reply Makefile.am:1078]: certainly, Thanks for your feedback

File: tests/non-executable-files-with-executable-bit.test [hunk 5]
```diff
@@ -8,6 +8,8 @@ find * -executable -type f | \
 	grep -vE "*.\.sh|tools/emoji" | # Ignore helper scripts \
 	grep -vE "src/ucd-tools/tools/(.*\.py|mkencodingtable)" | # Ignore ucd-tools helper scripts \
 	grep -vE "tests/.libs|src/ucd-tools/tests/print(ucd|c)data(_cpp)?" | # Ignore ucd-tools test programs \
+	grep -vE "test/fuzzing/*.py" | #Ignore fuzzing python helper script
```
Comment (sthibaul) [inline tests/non-executable-files-with-executable-bit.test:11]: missing `s` :)
Also, these are regexps, not shell expansions, so it'd rather be `.*\.py`

File: Makefile.am [hunk 6]
```diff
@@ -379,32 +387,16 @@ phsource/phonemes.stamp: \
...
-	find espeak-ng-data/lang -type f | zip -@ $@
-	find espeak-ng-data/voices -type f | grep -vF "/mb/" | zip -@ $@
-
-android/res/raw/espeakdata_version: android/res/raw/espeakdata.zip
-	sha1sum $< | awk '{ print $$1 }' > $@
-
-espeakdata: \
-	android/res/raw/espeakdata.zip \
-	android/res/raw/espeakdata_version
+espeakdata: espeak-ng-data/phontab dictionaries
```
Comment (sthibaul) [inline Makefile.am:399]: These are spurious changes

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -167,3 +168,47 @@ jobs:
...
+    - name: Store the fuzzer config
+      if: ${{ failure() }}
+      uses: actions/upload-artifact@v2
+      with:
+        name: fuzzer-config-${{ matrix.arch }}.log
+        path: tests/fuzzer-config.log
+    - name: make
+      run: make -j
+    - name: Fuzz function syth_espeak()
+      run: FUZZ_VOICE=en tests/fuzzing/synth_fuzzer -seed=1 tests/fuzzing/CORPUS_DIR
```
Comment (sthibaul) [inline .github/workflows/ci.yml:208]: You need to set the `ESPEAK_DATA_PATH` environment variable, see that CI https://github.com/yocvito/espeak-ng/runs/6248639305?check_suite_focus=true fails otherwise

File: Makefile.am [hunk 8]
```diff
@@ -333,9 +335,15 @@ tests_readclause_test_SOURCES += tests/dummy.cpp
 tests_ssml_fuzzer_test_SOURCES += tests/dummy.cpp
 endif
 
+
 if HAVE_LIBFUZZER
 tests_ssml_fuzzer_test_CFLAGS += -fsanitize=fuzzer
 tests_ssml_fuzzer_test_LDFLAGS = -fsanitize=fuzzer
+tests_fuzzing_synth_fuzzer_SOURCES = tests/fuzzing/synth_fuzzer.c
+tests_fuzzing_synth_fuzzer_LDADD =  src/libespeak-ng.la
+tests_fuzzing_synth_fuzzer_CFLAGS=  -fsanitize=address,undefined,fuzzer -fno-omit-frame-pointer -DPATH_ESPEAK_DATA=\"$(DATADIR)\" -g -fsanitize-coverage=edge -Isrc/libespeak-ng -g -O2 -fdebug-prefix-map=/root=. -fstack-protector-strong -Wformat -Werror=format-security -O1  -g -ggdb3  -fprofile-instr-generate -fcoverage-mapping ${AM_CFLAGS}
```
Comment (sthibaul) [inline Makefile.am:344]: Rather use `-DPATH_ESPEAK_DATA=\"$(abs_top_srcdir)/espeak-ng-data\"

**Comment by sthibaul** [comment]
Ok, so this now looks in good shape.

But... it finds bugs :)

Perhaps we can change the workflow rules to run this every week or so for now, so people get to know the bugs which need to be fixed before we can run this on master & pull requests?

**Comment by alex19EP** [comment]
> Perhaps we can change the workflow rules to run this every week or so for now, so people get to know the bugs which need to be fixed before we can run this on master & pull requests?

agree.
